### ISSUE TITLE
Flow: Memory leak in logging components

### ISSUE THREAD

**Issue by akselleirv** [grafana/agent#3728]
Hello,

I'm most likely having a memory leak when collecting logs using the flow mode. 

<img width="1646" alt="image" src="https://user-images.githubusercontent.com/52233080/235360840-cfd2db66-7e1b-40c3-936f-6185ce04e073.png">

I manage to trigger the issue when creating lots of pods which are restarted periodically by using the manifest provided blow: 

<details>
  <summary>Manifest example</summary>
  
```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: log-deployment
  namespace: logger
spec:
  replicas: 50
  selector:
    matchLabels:
      app: log-app
  template:
    metadata:
      labels:
        app: log-app
    spec:
      containers:
      - name: logger
        image: busybox
        args:
        - /bin/sh
        - -c
        - |
          i=0
          while [ $i -lt 10 ]; do
            echo "hello world $i"; 
            sleep 1; 
            i=$((i+1)); 
          done
          exit 0

```

</details>

<details>
  <summary>Flow config</summary>

```
discovery.kubernetes "pods" {
	role = "pod"
}

discovery.relabel "pods" {
	targets = discovery.kubernetes.pods.targets

	rule {
		source_labels = ["__meta_kubernetes_namespace"]
		target_label  = "namespace"
	}

	rule {
		source_labels = ["__meta_kubernetes_pod_container_name"]
		target_label  = "container"
	}

	rule {
		source_labels = ["__meta_kubernetes_pod_name"]
		target_label  = "pod"
	}

	rule {
		source_labels = ["__meta_kubernetes_pod_label_systemNumber"]
		target_label  = "systemNumber"
	}

	rule {
		source_labels = ["__meta_kubernetes_pod_label_privateLogs"]
		target_label  = "privateLogs"
	}

	rule {
		source_labels = ["__meta_kubernetes_pod_uid", "__meta_kubernetes_pod_container_name"]
		target_label  = "__path__"
		separator     = "/"
		replacement   = "/var/log/pods/*$1/*.log"
	}
}

discovery.file "pods" {
	path_targets = discovery.relabel.pods.output
}

loki.source.file "pods" {
	targets    = discovery.file.pods.targets
	forward_to = [loki.process.cri.receiver]
}

loki.process "cri" {
	stage.cri { }

	stage.label_drop {
		values = ["filename"]
	}

	forward_to = [loki.process.jwt_sanitization.receiver]
}


loki.process "jwt_sanitization" {
    stage.match {
        pipeline_name = "jwt_sanitization"
        selector = "{stream=~\"stdout|stderr\"} |~ \"eyJ[A-Za-z0-9-_]+.eyJ[A-Za-z0-9-_]+.[A-Za-z0-9-_.+]*\""

        stage.regex {
            expression = "(?P<jwt>eyJ[A-Za-z0-9-_]+\\.eyJ[A-Za-z0-9-_]+\\.[A-Za-z0-9-_.+/]*)"
        }

        stage.metrics {
            metric.counter {
                name        = "jwt_total"
                action      = "inc"
                description = "total count of redacted JWTs"
                source      = "jwt"
            }
        }

        stage.replace {
            expression = "(?P<jwt>eyJ[A-Za-z0-9-_]+\\.eyJ[A-Za-z0-9-_]+\\.[A-Za-z0-9-_.+/]*)"
            replace    = "<REDACTED_JWT>"
        }
	}

    forward_to = [loki.process.privatelogs.receiver]
}

loki.process "privatelogs" {
	stage.match {
		selector = "{privateLogs=\"enabled\"}"

		stage.tenant {
			label = "systemNumber"
		}
	}

	stage.match {
		selector = "{privateLogs!=\"enabled\"}"

		stage.tenant {
			value = "p-grp-aks-developers"
		}

		stage.label_drop {
			values = ["privateLogs"]
		}
	}

	forward_to = [loki.write.endpoint.receiver]
}


loki.write "endpoint" {
	external_labels = {
		cluster = "test",
	}

	endpoint {
		url               = "http://loki-gateway.loki.svc/loki/api/v1/push"
		bearer_token_file = "/run/secrets/remote-write/token"
	}
}
```
</details>

I've attached a zip file containing goroutines, allocs and heap pprofs: [debug.zip](https://github.com/grafana/agent/files/11361603/debug.zip)

**Comment by rfratto** [comment]
Thanks for reporting. Until we're able to reproduce, this one might take a while for us to track down. 

The heap profile is only showing 82MB of memory, interestingly. What does the `go_memstats_heap_inuse_bytes` metric report for the agent pod? 

Are you seeing any other leaks for other metrics from the agent, like `go_goroutines`?

**Comment by akselleirv** [comment]
Thank you the quick response. I'm using version v.0.33.0. 

The `go_memstats_heap_inuse_bytes` metric:
<img width="1404" alt="image" src="https://user-images.githubusercontent.com/52233080/235367876-884502b3-0809-4c07-832c-0b4deb85cdeb.png">


The `go_goroutines ` metric:
<img width="1389" alt="image" src="https://user-images.githubusercontent.com/52233080/235367792-cd1b04ee-1373-4e4a-bda3-28a7abbff2d2.png">


I'm able to replicate the issue using this script: 

```
#!/bin/bash

i=0
NUMBER_OF_CONCURRENT_PODS=20

while true; do
  for j in $(seq 1 $NUMBER_OF_CONCURRENT_PODS); do
    i=$((i+1))
    kubectl apply -f - <<EOF
apiVersion: v1
kind: Pod
metadata:
  name: logger-${i}
  labels:
    app: log-app
spec:
  restartPolicy: Never
  containers:
  - name: logger
    image: busybox:1.35.0
    args:
    - /bin/sh
    - -c
    - |
      n=0
      while [ \$n -lt 5 ]; do
        echo "hello world \$n"
        sleep 1
        n=\$((n+1))
      done
      exit 0
EOF
  done
  kubectl delete pods --field-selector=status.phase==Succeeded -l app=log-app
  sleep 20
done
```

This is 10 minutes of the script running and you can see a fairly steady memory usage increase:
<img width="1041" alt="image" src="https://user-images.githubusercontent.com/52233080/235368046-245c6fe2-49ea-4642-b50d-6898c067801a.png">

In the Agent logs it's also logging that it will start the tail routine for pods that already have existed for a while. I assume it will only start tailing on the pod creation and not start the tail routine every time the graph is evaluated? 

The Agent logs: [agent.log.zip](https://github.com/grafana/agent/files/11361950/agent.log.zip)

**Comment by rfratto** [comment]
Thank you, that made it very easy to reproduce the issue. I'm looking into it; this is definitely leaking somewhere:

```
github.com/grafana/agent/component/common/loki.NewEntryMutatorHandler.func1+0xa7
```

**Comment by rfratto** [comment]
Thanks, I found the leak and opened a fix at #3729. 

I was able to reproduce it without Kubernetes by modifying the script and config a little:

<details>
<summary>Bash script</summary>

```bash
#!/usr/bin/env bash

i=0
NUMBER_OF_CONCURRENT_FILES=20

LOG_FOLDER="$HOME/Desktop/repr/files"

rm -rf "$LOG_FOLDER/file-*.txt"

while true; do
  for old in $(seq 1 $i); do
    echo "Deleting $LOG_FOLDER/file-${old}.txt"
    rm -f "$LOG_FOLDER/file-${old}.txt"
  done

  for j in $(seq 1 $NUMBER_OF_CONCURRENT_FILES); do
    i=$((i+1))

    echo "Creating $LOG_FOLDER/file-${i}.txt"
    echo "hello, world!" >> "$LOG_FOLDER/file-${i}.txt"
  done

  sleep 5
done
```

</details>

<details>
<summary>Flow config</summary>

```river
discovery.file "files" {
	path_targets = [{
		__path__ = env("HOME") + "/Desktop/repr/files/*.txt",
	}]
}

loki.source.file "files" {
	targets    = discovery.file.files.targets
	forward_to = []
}
```

</details>

**Comment by rfratto** [comment]
> In the Agent logs it's also logging that it will start the tail routine for pods that already have existed for a while. I assume it will only start tailing on the pod creation and not start the tail routine every time the graph is evaluated?

With the way loki.source.file currently works, it restarts the tailers every time the component is updated, although it starts the new tailer at the position which the old tailer left off at. 

There's a TODO comment to optimize this behavior: https://github.com/grafana/agent/blob/4250181ad66901b16d46fd4f12bf96ebb98ac160/component/loki/source/file/file.go#L147-L149

**Comment by akselleirv** [comment]
Great, thank you for the quick fix. Would be appreciate if this fix was released in a v0.33.1.

**Comment by spartan0x117** [comment]
@akselleirv I'm planning on including this change in `v0.33.1`
